service/src/main/java/bio/terra/pipelines/stairway/FetchQuotaConsumedFromDataTableStep.java

```diff
@@ -0,0 +1,93 @@
+import bio.terra.rawls.model.Entity;
+import bio.terra.stairway.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This step calls Rawls to fetch outputs from a data table row for a given quota consumed job. It
+ * specifically fetches the quota consumed value from the data table row using the quota_consumed
+ * key
+ *
+ * <p>This step expects nothign from the working map
+ */
+public class FetchQuotaConsumedFromDataTableStep implements Step {
+
+  private final RawlsService rawlsService;
+  private final SamService samService;
+  private final Logger logger = LoggerFactory.getLogger(FetchQuotaConsumedFromDataTableStep.class);
+
+  public FetchQuotaConsumedFromDataTableStep(RawlsService rawlsService, SamService samService) {
+    this.rawlsService = rawlsService;
+    this.samService = samService;
+  }
+
+  @Override
+  @SuppressWarnings(
+      "java:S2259") // suppress warning for possible NPE when calling pipelineName.getValue(),
+  //  since we do validate that pipelineName is not null in `validateRequiredEntries`
+  public StepResult doStep(FlightContext flightContext) {
+    String jobId = flightContext.getFlightId();
+
+    // validate and extract parameters from input map
+    var inputParameters = flightContext.getInputParameters();
+    FlightUtils.validateRequiredEntries(
+        inputParameters,
+        JobMapKeys.PIPELINE_NAME,
+        ImputationJobMapKeys.CONTROL_WORKSPACE_BILLING_PROJECT,
+        ImputationJobMapKeys.CONTROL_WORKSPACE_NAME);
+
+    String controlWorkspaceBillingProject =
+        inputParameters.get(ImputationJobMapKeys.CONTROL_WORKSPACE_BILLING_PROJECT, String.class);
+    String controlWorkspaceName =
+        inputParameters.get(ImputationJobMapKeys.CONTROL_WORKSPACE_NAME, String.class);
+    PipelinesEnum pipelineName = inputParameters.get(JobMapKeys.PIPELINE_NAME, PipelinesEnum.class);
+
+    Entity entity;
+    try {
+      entity =
+          rawlsService.getDataTableEntity(
+              samService.getTeaspoonsServiceAccountToken(),
+              controlWorkspaceBillingProject,
+              controlWorkspaceName,
+              pipelineName.getValue(),
+              jobId);
+    } catch (RawlsServiceApiException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+
+    // extract quota_consumed from entity
+    int quotaConsumed;
+    try {
+      quotaConsumed = (int) entity.getAttributes().get("quota_consumed");
+      if (quotaConsumed <= 0) {
+        return new StepResult(
+            StepStatus.STEP_RESULT_FAILURE_FATAL,
+            new InternalServerErrorException("Quota consumed is unexpectedly not greater than 0"));
+      }
+    } catch (NullPointerException e) {
+      return new StepResult(
+          StepStatus.STEP_RESULT_FAILURE_FATAL,
+          new InternalServerErrorException("Quota consumed is unexpectedly null"));
+    }
+
+    logger.info("Quota consumed: {}", quotaConsumed);
```

Comment:
is the idea that in future we'll store this in the working map to be used by the logic to do the quota check in a subsequent step?

Reply:
i think we may do the quota check here or in a subsequent step - thats due to be done in https://broadworkbench.atlassian.net/browse/TSPS-361.  I'd prob do it here and rename the step